File: test-listing-flow.py
"""
Test Script for VDI Realty FSBO Listing Flow (Python Version)

This script tests the complete listing lifecycle:
1. Submit a new FSBO listing  
2. Verify it appears in public listings (fsbo-listings.html, listings.html)
3. Verify it appears in admin panel
4. Delete the listing via admin
5. Verify it's removed from public view but still in admin with 'removed' status

Usage: python test-listing-flow.py

Requirements: Python 3.6+, requests library (pip install requests)
"""

import requests
import json
import time
from datetime import datetime
import sys

API_BASE_URL = 'https://api.vdirealty.com'
ADMIN_PASSWORD = 'VDI2025AdminSecure789'  # From .env

# ANSI color codes for terminal output
class Colors:
    RESET = '\033[0m'
    GREEN = '\033[32m'
    RED = '\033[31m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    CYAN = '\033[36m'
    BOLD = '\033[1m'

def log(message, color='RESET'):
    color_code = getattr(Colors, color, Colors.RESET)
    print(f"{color_code}{message}{Colors.RESET}")

def log_step(step, description):
    log(f"\n{'=' * 60}", 'CYAN')
    log(f"STEP {step}: {description}", 'BOLD')
    log('=' * 60, 'CYAN')

def log_success(message):
    log(f"✓ {message}", 'GREEN')

def log_error(message):
    log(f"✗ {message}", 'RED')

def log_info(message):
    log(f"ℹ {message}", 'BLUE')

# Test data for a residential listing
test_listing = {
    # Contact Info
    'firstName': 'Test',
    'lastName': 'Seller',
    'email': 'test.seller@example.com',
    'phone': '(206) 555-0123',
    
    # Property Address
    'address': '123 Test Street',
    'city': 'Seattle',
    'state': 'WA',
    'zip': '98101',
    
    # Property Type
    'propertyType': 'Single Family',
    
    # Property Details
    'price': 650000,
    'sqft': 2400,
    'bedrooms': 3,
    'bathrooms': 2.5,
    'yearBuilt': 2015,
    'lotSize': 0.25,
    
    # Features & Description
    'features': 'Hardwood floors, Updated kitchen, Master suite, Garage, Fenced yard',
    'description': 'Beautiful test property in the heart of Seattle. This is a test listing created by the automated test script. Features modern finishes, open floor plan, and great location. Perfect for testing the VDI Realty FSBO listing system.',
    
    # Privacy
    'privateContact': False,
    
    # MLS/External (optional)
    'mlsNumber': '',
    'externalUrl': ''
}

def api_request(method, endpoint, data=None, headers=None):
    """Make an API request and return the response"""
    url = f"{API_BASE_URL}{endpoint}"
    headers = headers or {}
    headers['Content-Type'] = 'application/json'
    
    try:
        if method == 'GET':
            response = requests.get(url, headers=headers, timeout=10)
        elif method == 'POST':
            response = requests.post(url, json=data, headers=headers, timeout=10)
        elif method == 'DELETE':
            response = requests.delete(url, headers=headers, timeout=10)
        else:
            raise ValueError(f"Unsupported method: {method}")
        
        return {
            'ok': response.ok,
            'status': response.status_code,
            'data': response.json() if response.text else {}
        }
    except requests.exceptions.RequestException as e:
        return {
            'ok': False,
            'status': 0,
            'error': str(e)
        }

def test_submit_listing():
    """Test Step 1: Submit a new FSBO listing"""
    log_step(1, 'Submit New FSBO Listing')
    
    log_info(f"Submitting test listing for: {test_listing['address']}, {test_listing['city']}")
    log_info(f"Property Type: {test_listing['propertyType']}")
    log_info(f"Price: ${test_listing['price']:,}")
    log_info(f"Details: {test_listing['bedrooms']} bed, {test_listing['bathrooms']} bath, {test_listing['sqft']} sqft")
    
    result = api_request('POST', '/api/fsbo/submit', test_listing)
    
    if result['ok'] and result['data'].get('success'):
        log_success("Listing created successfully!")
        listing_id = result['data']['listingId']
        expiration = result['data']['expirationDate']
        log_info(f"Listing ID: {listing_id}")
        log_info(f"Expiration Date: {datetime.fromisoformat(expiration.replace('Z', '+00:00')).strftime('%Y-%m-%d')}")
        return listing_id
    else:
        error_msg = result['data'].get('error') or result.get('error') or 'Unknown error'
        log_error(f"Failed to create listing: {error_msg}")
        raise Exception('Failed to create listing')

def test_public_listings(expected_listing_id):
    """Test Step 2: Verify listing appears in public view"""
    log_step(2, 'Verify Listing Appears in Public View')
    
    log_info('Fetching public listings (GET /api/fsbo/listings)...')
    
    result = api_request('GET', '/api/fsbo/listings')
    
    if result['ok'] and result['data'].get('success'):
        count = result['data']['count']
        log_success(f"Retrieved {count} public listing(s)")
        
        # Find our test listing
        listings = result['data']['listings']
        our_listing = next((l for l in listings if l['id'] == expected_listing_id), None)
        
        if our_listing:
            log_success("Test listing found in public listings!")
            log_info(f"Address: {our_listing['address']}, {our_listing['city']}, {our_listing['state']} {our_listing['zip']}")
            log_info(f"Price: ${our_listing['price']:,}")
            log_info(f"Status: {our_listing['status']}")
            log_info(f"Photos: {len(our_listing['photos'])} attached")
            
            # Check contact visibility
            if our_listing.get('privateContact'):
                log_info("Contact: Hidden (private listing)")
            else:
                log_info(f"Contact: {our_listing.get('email')}, {our_listing.get('phone')}")
            
            return True
        else:
            log_error("Test listing NOT found in public listings!")
            log_info(f"Available listing IDs: {', '.join(str(l['id']) for l in listings)}")
            return False
    else:
        error_msg = result['data'].get('error') or result.get('error')
        log_error(f"Failed to fetch public listings: {error_msg}")
        return False

def test_admin_listings(expected_listing_id, expected_status='active'):
    """Test Step 3: Verify listing appears in admin panel"""
    log_step(3, 'Verify Listing Appears in Admin Panel')
    
    log_info('Fetching admin listings (requires authentication)...')
    
    result = api_request('GET', '/api/fsbo/admin/listings', headers={
        'Authorization': ADMIN_PASSWORD
    })
    
    if result['ok'] and result['data'].get('success'):
        listings = result['data']['listings']
        log_success(f"Retrieved {len(listings)} total listing(s) from admin panel")
        
        # Find our test listing
        our_listing = next((l for l in listings if l['id'] == expected_listing_id), None)
        
        if our_listing:
            log_success("Test listing found in admin panel!")
            log_info(f"ID: {our_listing['id']}")
            log_info(f"Address: {our_listing['address']}, {our_listing['city']}")
            log_info(f"Status: {our_listing['status']}")
            log_info(f"Contact: {our_listing['firstName']} {our_listing['lastName']}")
            log_info(f"Email: {our_listing['email']}")
            log_info(f"Phone: {our_listing['phone']}")
            log_info(f"Submitted: {datetime.fromisoformat(our_listing['submissionDate'].replace('Z', '+00:00')).strftime('%Y-%m-%d %H:%M')}")
            log_info(f"Expires: {datetime.fromisoformat(our_listing['expirationDate'].replace('Z', '+00:00')).strftime('%Y-%m-%d %H:%M')}")
            
            if our_listing['status'] == expected_status:
                log_success(f'Status matches expected: "{expected_status}"')
                return True
            else:
                log_error(f'Status mismatch! Expected "{expected_status}", got "{our_listing["status"]}"')
                return False
        else:
            log_error("Test listing NOT found in admin panel!")
            log_info(f"Available listing IDs: {', '.join(str(l['id']) for l in listings)}")
            return False
    elif result['status'] == 401:
        log_error("Authentication failed! Admin password is incorrect.")
        return False
    else:
        error_msg = result['data'].get('error') or result.get('error')
        log_error(f"Failed to fetch admin listings: {error_msg}")
        return False

def test_delete_listing(listing_id):
    """Test Step 4: Delete listing via admin panel"""
    log_step(4, 'Delete Listing via Admin Panel')
    
    log_info(f"Deleting listing ID: {listing_id}...")
    
    result = api_request('DELETE', f'/api/fsbo/listing/{listing_id}', headers={
        'Authorization': ADMIN_PASSWORD
    })
    
    if result['ok'] and result['data'].get('success'):
        log_success("Listing deleted successfully!")
        log_info(result['data']['message'])
        return True
    elif result['status'] == 401:
        log_error("Authentication failed! Cannot delete without admin password.")
        return False
    else:
        error_msg = result['data'].get('error') or result.get('error')
        log_error(f"Failed to delete listing: {error_msg}")
        return False

def test_verify_deletion(listing_id):
    """Test Step 5: Verify listing removed from public view"""
    log_step(5, 'Verify Listing Removed from Public View')
    
    log_info('Checking public listings...')
    
    public_result = api_request('GET', '/api/fsbo/listings')
    
    if public_result['ok'] and public_result['data'].get('success'):
        listings = public_result['data']['listings']
        found_in_public = next((l for l in listings if l['id'] == listing_id), None)
        
        if not found_in_public:
            log_success("Listing correctly removed from public view!")
        else:
            log_error("Listing still appears in public listings!")
            return False
    else:
        log_error("Failed to fetch public listings for verification")
        return False
    
    log_info('Checking admin panel...')
    
    admin_result = api_request('GET', '/api/fsbo/admin/listings', headers={
        'Authorization': ADMIN_PASSWORD
    })
    
    if admin_result['ok'] and admin_result['data'].get('success'):
        listings = admin_result['data']['listings']
        admin_listing = next((l for l in listings if l['id'] == listing_id), None)
        
        if admin_listing:
            if admin_listing['status'] == 'removed':
                log_success('Listing still visible in admin panel with status "removed"!')
                log_info("This is correct - soft delete preserves the record.")
                return True
            else:
                log_error(f'Listing found in admin but status is "{admin_listing["status"]}" instead of "removed"')
                return False
        else:
            log_error('Listing not found in admin panel! It should be there with status "removed".')
            return False
    else:
        log_error("Failed to fetch admin listings for verification")
        return False

def run_tests():
    """Main test runner"""
    log('\n' + '═' * 60, 'BOLD')
    log('  VDI REALTY FSBO LISTING SYSTEM - INTEGRATION TEST  ', 'BOLD')
    log('═' * 60 + '\n', 'BOLD')
    
    log(f"API Endpoint: {API_BASE_URL}", 'CYAN')
    log(f"Test Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}", 'CYAN')
    log(f"Admin Auth: {'Configured ✓' if ADMIN_PASSWORD else 'Missing ✗'}", 'CYAN')
    
    tests_passed = 0
    tests_failed = 0
    listing_id = None
    
    try:
        # Test 1: Submit listing
        listing_id = test_submit_listing()
        tests_passed += 1
        
        # Wait a moment for database to settle
        time.sleep(1)
        
        # Test 2: Verify in public listings
        if test_public_listings(listing_id):
            tests_passed += 1
        else:
            tests_failed += 1
        
        # Test 3: Verify in admin panel
        if test_admin_listings(listing_id, 'active'):
            tests_passed += 1
        else:
            tests_failed += 1
        
        # Test 4: Delete listing
        if test_delete_listing(listing_id):
            tests_passed += 1
        else:
            tests_failed += 1
            raise Exception('Cannot continue without successful deletion')
        
        # Wait a moment for deletion to process
        time.sleep(1)
        
        # Test 5: Verify deletion
        if test_verify_deletion(listing_id):
            tests_passed += 1
        else:
            tests_failed += 1
        
    except Exception as e:
        log_error(f"\nTest suite failed: {str(e)}")
        tests_failed += 1
    
    # Final summary
    log('\n' + '═' * 60, 'BOLD')
    log('  TEST SUMMARY  ', 'BOLD')
    log('═' * 60, 'BOLD')
    log(f"Total Tests: {tests_passed + tests_failed}", 'CYAN')
    log(f"Passed: {tests_passed}", 'GREEN')
    log(f"Failed: {tests_failed}", 'RED')
    
    if listing_id:
        log(f"\nTest Listing ID: {listing_id}", 'CYAN')
        log(f"View in Admin Panel: https://www.vdirealty.com/admin.html", 'CYAN')
        log(f"Password: {ADMIN_PASSWORD}", 'CYAN')
    
    if tests_failed == 0:
        log('\n🎉 ALL TESTS PASSED! ✓', 'GREEN')
        log('The FSBO listing system is working correctly.\n', 'GREEN')
        return 0
    else:
        log('\n❌ SOME TESTS FAILED', 'RED')
        log('Please review the errors above and check the system.\n', 'RED')
        return 1

if __name__ == '__main__':
    # Check if requests library is available
    try:
        import requests
    except ImportError:
        print("Error: 'requests' library not found.")
        print("Please install it with: pip install requests")
        sys.exit(1)
    
    exit_code = run_tests()
    sys.exit(exit_code)

File: test-listing-flow.js
/**
 * Test Script for VDI Realty FSBO Listing Flow
 * 
 * This script tests the complete listing lifecycle:
 * 1. Submit a new FSBO listing
 * 2. Verify it appears in public listings (fsbo-listings.html, listings.html)
 * 3. Verify it appears in admin panel
 * 4. Delete the listing via admin
 * 5. Verify it's removed from public view but still in admin with 'removed' status
 * 
 * Usage: node test-listing-flow.js
 */

const API_BASE_URL = 'https://api.vdirealty.com';
const ADMIN_PASSWORD = 'VDI2025AdminSecure789'; // From .env

// ANSI color codes for terminal output
const colors = {
    reset: '\x1b[0m',
    green: '\x1b[32m',
    red: '\x1b[31m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    cyan: '\x1b[36m',
    bold: '\x1b[1m'
};

function log(message, color = 'reset') {
    console.log(`${colors[color]}${message}${colors.reset}`);
}

function logStep(step, description) {
    log(`\n${'='.repeat(60)}`, 'cyan');
    log(`STEP ${step}: ${description}`, 'bold');
    log('='.repeat(60), 'cyan');
}

function logSuccess(message) {
    log(`✓ ${message}`, 'green');
}

function logError(message) {
    log(`✗ ${message}`, 'red');
}

function logInfo(message) {
    log(`ℹ ${message}`, 'blue');
}

// Test data for a residential listing
const testListing = {
    // Contact Info
    firstName: 'Test',
    lastName: 'Seller',
    email: 'test.seller@example.com',
    phone: '(206) 555-0123',
    
    // Property Address
    address: '123 Test Street',
    city: 'Seattle',
    state: 'WA',
    zip: '98101',
    
    // Property Type
    propertyType: 'Single Family',
    
    // Property Details
    price: 650000,
    sqft: 2400,
    bedrooms: 3,
    bathrooms: 2.5,
    yearBuilt: 2015,
    lotSize: 0.25,
    
    // Features & Description
    features: 'Hardwood floors, Updated kitchen, Master suite, Garage, Fenced yard',
    description: 'Beautiful test property in the heart of Seattle. This is a test listing created by the automated test script. Features modern finishes, open floor plan, and great location. Perfect for testing the VDI Realty FSBO listing system.',
    
    // Privacy
    privateContact: false,
    
    // MLS/External (optional)
    mlsNumber: '',
    externalUrl: ''
};

// Sleep helper function
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// API helper functions
async function apiRequest(method, endpoint, data = null, headers = {}) {
    const url = `${API_BASE_URL}${endpoint}`;
    const options = {
        method,
        headers: {
            'Content-Type': 'application/json',
            ...headers
        }
    };
    
    if (data && method !== 'GET') {
        options.body = JSON.stringify(data);
    }
    
    try {
        const response = await fetch(url, options);
        const responseData = await response.json();
        
        return {
            ok: response.ok,
            status: response.status,
            data: responseData
        };
    } catch (error) {
        return {
            ok: false,
            status: 0,
            error: error.message
        };
    }
}

// Test functions
async function testSubmitListing() {
    logStep(1, 'Submit New FSBO Listing');
    
    logInfo(`Submitting test listing for: ${testListing.address}, ${testListing.city}`);
    logInfo(`Property Type: ${testListing.propertyType}`);
    logInfo(`Price: $${testListing.price.toLocaleString()}`);
    logInfo(`Details: ${testListing.bedrooms} bed, ${testListing.bathrooms} bath, ${testListing.sqft} sqft`);
    
    const result = await apiRequest('POST', '/api/fsbo/submit', testListing);
    
    if (result.ok && result.data.success) {
        logSuccess(`Listing created successfully!`);
        logInfo(`Listing ID: ${result.data.listingId}`);
        logInfo(`Expiration Date: ${new Date(result.data.expirationDate).toLocaleDateString()}`);
        return result.data.listingId;
    } else {
        logError(`Failed to create listing: ${result.data.error || result.error || 'Unknown error'}`);
        throw new Error('Failed to create listing');
    }
}

async function testPublicListings(expectedListingId) {
    logStep(2, 'Verify Listing Appears in Public View');
    
    logInfo('Fetching public listings (GET /api/fsbo/listings)...');
    
    const result = await apiRequest('GET', '/api/fsbo/listings');
    
    if (result.ok && result.data.success) {
        logSuccess(`Retrieved ${result.data.count} public listing(s)`);
        
        // Find our test listing
        const ourListing = result.data.listings.find(l => l.id === expectedListingId);
        
        if (ourListing) {
            logSuccess(`Test listing found in public listings!`);
            logInfo(`Address: ${ourListing.address}, ${ourListing.city}, ${ourListing.state} ${ourListing.zip}`);
            logInfo(`Price: $${ourListing.price.toLocaleString()}`);
            logInfo(`Status: ${ourListing.status}`);
            logInfo(`Photos: ${ourListing.photos.length} attached`);
            
            // Check contact visibility
            if (ourListing.privateContact) {
                logInfo(`Contact: Hidden (private listing)`);
            } else {
                logInfo(`Contact: ${ourListing.email}, ${ourListing.phone}`);
            }
            
            return true;
        } else {
            logError(`Test listing NOT found in public listings!`);
            logInfo(`Available listing IDs: ${result.data.listings.map(l => l.id).join(', ')}`);
            return false;
        }
    } else {
        logError(`Failed to fetch public listings: ${result.data.error || result.error}`);
        return false;
    }
}

async function testAdminListings(expectedListingId, expectedStatus = 'active') {
    logStep(3, 'Verify Listing Appears in Admin Panel');
    
    logInfo('Fetching admin listings (requires authentication)...');
    
    const result = await apiRequest('GET', '/api/fsbo/admin/listings', null, {
        'Authorization': ADMIN_PASSWORD
    });
    
    if (result.ok && result.data.success) {
        logSuccess(`Retrieved ${result.data.listings.length} total listing(s) from admin panel`);
        
        // Find our test listing
        const ourListing = result.data.listings.find(l => l.id === expectedListingId);
        
        if (ourListing) {
            logSuccess(`Test listing found in admin panel!`);
            logInfo(`ID: ${ourListing.id}`);
            logInfo(`Address: ${ourListing.address}, ${ourListing.city}`);
            logInfo(`Status: ${ourListing.status}`);
            logInfo(`Contact: ${ourListing.firstName} ${ourListing.lastName}`);
            logInfo(`Email: ${ourListing.email}`);
            logInfo(`Phone: ${ourListing.phone}`);
            logInfo(`Submitted: ${new Date(ourListing.submissionDate).toLocaleString()}`);
            logInfo(`Expires: ${new Date(ourListing.expirationDate).toLocaleString()}`);
            
            if (ourListing.status === expectedStatus) {
                logSuccess(`Status matches expected: "${expectedStatus}"`);
                return true;
            } else {
                logError(`Status mismatch! Expected "${expectedStatus}", got "${ourListing.status}"`);
                return false;
            }
        } else {
            logError(`Test listing NOT found in admin panel!`);
            logInfo(`Available listing IDs: ${result.data.listings.map(l => l.id).join(', ')}`);
            return false;
        }
    } else if (result.status === 401) {
        logError(`Authentication failed! Admin password is incorrect.`);
        return false;
    } else {
        logError(`Failed to fetch admin listings: ${result.data.error || result.error}`);
        return false;
    }
}

async function testDeleteListing(listingId) {
    logStep(4, 'Delete Listing via Admin Panel');
    
    logInfo(`Deleting listing ID: ${listingId}...`);
    
    const result = await apiRequest('DELETE', `/api/fsbo/listing/${listingId}`, null, {
        'Authorization': ADMIN_PASSWORD
    });
    
    if (result.ok && result.data.success) {
        logSuccess(`Listing deleted successfully!`);
        logInfo(result.data.message);
        return true;
    } else if (result.status === 401) {
        logError(`Authentication failed! Cannot delete without admin password.`);
        return false;
    } else {
        logError(`Failed to delete listing: ${result.data.error || result.error}`);
        return false;
    }
}

async function testVerifyDeletion(listingId) {
    logStep(5, 'Verify Listing Removed from Public View');
    
    logInfo('Checking public listings...');
    
    const publicResult = await apiRequest('GET', '/api/fsbo/listings');
    
    if (publicResult.ok && publicResult.data.success) {
        const foundInPublic = publicResult.data.listings.find(l => l.id === listingId);
        
        if (!foundInPublic) {
            logSuccess(`Listing correctly removed from public view!`);
        } else {
            logError(`Listing still appears in public listings!`);
            return false;
        }
    } else {
        logError(`Failed to fetch public listings for verification`);
        return false;
    }
    
    logInfo('Checking admin panel...');
    
    const adminResult = await apiRequest('GET', '/api/fsbo/admin/listings', null, {
        'Authorization': ADMIN_PASSWORD
    });
    
    if (adminResult.ok && adminResult.data.success) {
        const adminListing = adminResult.data.listings.find(l => l.id === listingId);
        
        if (adminListing) {
            if (adminListing.status === 'removed') {
                logSuccess(`Listing still visible in admin panel with status "removed"!`);
                logInfo(`This is correct - soft delete preserves the record.`);
                return true;
            } else {
                logError(`Listing found in admin but status is "${adminListing.status}" instead of "removed"`);
                return false;
            }
        } else {
            logError(`Listing not found in admin panel! It should be there with status "removed".`);
            return false;
        }
    } else {
        logError(`Failed to fetch admin listings for verification`);
        return false;
    }
}

// Main test runner
async function runTests() {
    log('\n' + '═'.repeat(60), 'bold');
    log('  VDI REALTY FSBO LISTING SYSTEM - INTEGRATION TEST  ', 'bold');
    log('═'.repeat(60) + '\n', 'bold');
    
    log(`API Endpoint: ${API_BASE_URL}`, 'cyan');
    log(`Test Time: ${new Date().toLocaleString()}`, 'cyan');
    log(`Admin Auth: ${ADMIN_PASSWORD ? 'Configured ✓' : 'Missing ✗'}`, 'cyan');
    
    let testsPassed = 0;
    let testsFailed = 0;
    let listingId = null;
    
    try {
        // Test 1: Submit listing
        listingId = await testSubmitListing();
        testsPassed++;
        
        // Wait a moment for database to settle
        await sleep(1000);
        
        // Test 2: Verify in public listings
        if (await testPublicListings(listingId)) {
            testsPassed++;
        } else {
            testsFailed++;
        }
        
        // Test 3: Verify in admin panel
        if (await testAdminListings(listingId, 'active')) {
            testsPassed++;
        } else {
            testsFailed++;
        }
        
        // Test 4: Delete listing
        if (await testDeleteListing(listingId)) {
            testsPassed++;
        } else {
            testsFailed++;
            throw new Error('Cannot continue without successful deletion');
        }
        
        // Wait a moment for deletion to process
        await sleep(1000);
        
        // Test 5: Verify deletion
        if (await testVerifyDeletion(listingId)) {
            testsPassed++;
        } else {
            testsFailed++;
        }
        
    } catch (error) {
        logError(`\nTest suite failed: ${error.message}`);
        testsFailed++;
    }
    
    // Final summary
    log('\n' + '═'.repeat(60), 'bold');
    log('  TEST SUMMARY  ', 'bold');
    log('═'.repeat(60), 'bold');
    log(`Total Tests: ${testsPassed + testsFailed}`, 'cyan');
    log(`Passed: ${testsPassed}`, 'green');
    log(`Failed: ${testsFailed}`, 'red');
    
    if (listingId) {
        log(`\nTest Listing ID: ${listingId}`, 'cyan');
        log(`View in Admin Panel: https://www.vdirealty.com/admin.html`, 'cyan');
        log(`Password: ${ADMIN_PASSWORD}`, 'cyan');
    }
    
    if (testsFailed === 0) {
        log('\n🎉 ALL TESTS PASSED! ✓', 'green');
        log('The FSBO listing system is working correctly.\n', 'green');
        process.exit(0);
    } else {
        log('\n❌ SOME TESTS FAILED', 'red');
        log('Please review the errors above and check the system.\n', 'red');
        process.exit(1);
    }
}

// Run the tests
runTests().catch(error => {
    logError(`\nFatal error: ${error.message}`);
    console.error(error);
    process.exit(1);
});
